executor/src/subscriber.rs

```diff
@@ -120,144 +153,155 @@ impl Subscriber {
+        let mut ret = vec![];
+        let authorities = deliver.certificate.signed_authorities(&self.committee);
+        for (batch_index, (digest, worker_id)) in
+            deliver.certificate.header.payload.iter().enumerate()
+        {
+            let batch_index = BatchIndex {
+                consensus_output: deliver.clone(),
+                next_certificate_index: deliver.consensus_index,
+                batch_index: batch_index as u64,
+            };
+            ret.push(
+                self.fetch_payload(*digest, *worker_id, authorities.clone())
+                    .map(move |batch| (batch_index, batch)),
+            );
         }
 
-        Ok(())
+        ret
     }
 
-    /// Downloads the payload from the worker and forwards the output to the
-    /// executor channel if the operation is successful. An error is returned
-    /// if we can't forward the output to the executor. If an irrecoverable error
-    /// has occurred while downloading the payload then this method panics.
-    #[instrument(level="debug", skip_all, fields(certificate_id = ?message.certificate.digest()), err)]
-    async fn download_payload_and_forward(&self, message: ConsensusOutput) -> SubscriberResult<()> {
-        // Fetch the certificate's payload from the workers. This is done via the
-        // block_waiter component. If the batches are not available in the workers then
-        // block_waiter will do its best to sync from the other peers. Once all batches
-        // are available, we forward the certificate to the Executor Core.
-        let result = Self::wait_on_payload(
-            self.metrics.clone(),
-            self.get_block_retry_policy.clone(),
-            self.store.clone(),
-            self.tx_get_block_commands.clone(),
-            message,
-        )
-        .await;
-
-        match result {
-            Ok(output) => {
-                if self.tx_executor.send(output).await.is_err() {
-                    return Err(SubscriberError::ClosedChannel(
-                        stringify!(self.tx_executor).to_owned(),
-                    ));
+    /// Fetches single payload from network
+    /// This future performs infinite retries and blocks until Batch is available
+    /// As an optimization it tries to download from local worker first, but then fans out
+    /// requests to remote worker if not found locally
+    #[instrument(level = "debug", skip_all, fields(digest = % digest, worker_id = % worker_id))]
+    async fn fetch_payload(
+        &self,
+        digest: BatchDigest,
+        worker_id: WorkerId,
+        authorities: Vec<PublicKey>,
+    ) -> Batch {
+        if let Some(payload) = self.try_fetch_locally(digest, worker_id).await {
+            return payload;
+        }
+        let _timer = self.metrics.subscriber_remote_fetch_latency.start_timer();
+        let mut stagger = Duration::from_secs(0);
+        let mut futures = vec![];
+        for authority in &authorities {
```

Comment:
Do we need to randomize the order of authorities to avoid overloading the authority that appears first? Also the order of authority to query may be relevant depending on how we handle response from byzantine workers.